### PR TITLE
add max/min length rule

### PR DESCRIPTION
**Pull request by fkpwolf**
create new rule for description maxlength & minlength

Feature: https://ceapex.visualstudio.com/Engineering/_workitems/edit/166621
[AB#166621](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/166621)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5643)

### CODE REVIEW

File: src/docfx/config/ops/OpsMetadataRuleConverter.cs [hunk 1]
```diff
@@ -124,6 +126,7 @@ private static bool TryGetAttributeAdditionalErrors(Dictionary<string, OpsMetada
                 { "MicrosoftAlias", new string[] { "ms-alias-invalid" } },
                 { "Deprecated", new string[] { "attribute-deprecated" } },
                 { "Uniqueness", new string[] { "duplicate-attribute" } },
+                { "Length", new string[] { "description-bad-length" } },
```
Comment (yufeih) [inline src/docfx/config/ops/OpsMetadataRuleConverter.cs]: Map `Length` to a general error code that could apply to any attribute, so we can reuse it for checking length for other metadatas.

```suggestion
                { "Length", new string[] { "string-length-invalid" } },
```

File: src/docfx/config/ops/OpsMetadataRuleConverter.cs [hunk 2]
```diff
@@ -231,6 +234,26 @@ private static bool TryGetAllowlist(string parentName, string? listId, Allowlist
             return null;
         }
 
+        private static int? GetMinLength(Dictionary<string, OpsMetadataRule> rulesInfo)
+        {
+            if (rulesInfo.TryGetValue("Length", out var dateRuleInfo))
```
Comment (herohua) [inline src/docfx/config/ops/OpsMetadataRuleConverter.cs]: rename to `lengthRuleInfo`

File: test/docfx.Test/data/validation/schema.json [hunk 3]
```diff
@@ -5234,6 +5236,11 @@
         "severity": "suggestion",
         "code": "duplicate-description-within-docset",
         "additionalMessage": "Description must be unique within a docset."
+      },
+      "description-bad-length": {
+        "severity": "suggestion",
+        "code": "description-bad-length",
+        "additionalMessage": "Description length of characters is out of range. The description should be between MinLength and MaxLength characters, including spaces."
```
Comment (herohua) [inline test/docfx.Test/data/validation/schema.json]: would it be better that we include the actual min and max number in error message?

File: src/docfx/config/ops/OpsMetadataRuleConverter.cs [hunk 4]
```diff
@@ -231,6 +234,26 @@ private static bool TryGetAllowlist(string parentName, string? listId, Allowlist
...
+            {
+                return dateRuleInfo.MinLength;
+            }
+
+            return null;
+        }
+
+        private static int? GetMaxLength(Dictionary<string, OpsMetadataRule> rulesInfo)
+        {
+            if (rulesInfo.TryGetValue("Length", out var dateRuleInfo))
```
Comment (herohua) [inline src/docfx/config/ops/OpsMetadataRuleConverter.cs]: same here

File: test/docfx.Test/data/validation/schema.json (hunk 3 above)
Comment (yufeih) [reply test/docfx.Test/data/validation/schema.json]: That is already included in the first part of the message. This is `additionalMessage` that is appended to the general message.

File: src/docfx/config/ops/OpsMetadataRuleConverter.cs (hunk 1 above)
Comment (fkpwolf) [reply src/docfx/config/ops/OpsMetadataRuleConverter.cs]: ok. yml test case added.

File: src/docfx/config/ops/OpsMetadataRuleConverter.cs (hunk 2 above)
Comment (fkpwolf) [reply src/docfx/config/ops/OpsMetadataRuleConverter.cs]: ok

File: src/docfx/config/ops/OpsMetadataRuleConverter.cs (hunk 4 above)
Comment (fkpwolf) [reply src/docfx/config/ops/OpsMetadataRuleConverter.cs]: ok

File: test/docfx.Test/data/validation/schema.json (hunk 3 above)
Comment (fkpwolf) [reply test/docfx.Test/data/validation/schema.json]: the message can be found in yml test case.

**Comment by superyyrrzz** [comment]
AB#166621

**Comment by superyyrrzz** [comment]
AB#166621

File: docs/specs/validation.yml [hunk 5]
```diff
@@ -448,3 +448,37 @@ outputs:
     ["suggestion","duplicate-attribute","Attribute 'description' with value 'description1' is duplicated in 'a.md(3,14)', 'b.md(2,14)'","b.md",2,14]
     ["suggestion","duplicate-attribute","Attribute 'description' with value 'description1' is duplicated in 'a.md(3,14)', 'b.md(2,14)'","a.md",3,14]
 ---
+# Json schema length rule: the length of string must be in range of MinLength and MaxLength
+inputs:
+  docfx.yml: |
+    metadataSchema:
+      - schema.json
+  schema.json: |
```
Comment (yufeih) [inline docs/specs/validation.yml]: refer to test cases with `markdownValidationRules`, to cover the rule --> JsonSchema conversion logic

Comment (OsmondJiang) [reply docs/specs/validation.yml]: that's a metadata rule validation, not a content rule, @yufeih

**Pull request merged**